From 27601aabfd33d9d6fc54938d9df81de1fe1cadeb Mon Sep 17 00:00:00 2001
Message-Id: <27601aabfd33d9d6fc54938d9df81de1fe1cadeb.1431284209.git.chang-joon.lee@intel.com>
In-Reply-To: <89460f3fc3d6c0659d44897a6adec3c3fecc5564.1431284209.git.chang-joon.lee@intel.com>
References: <89460f3fc3d6c0659d44897a6adec3c3fecc5564.1431284209.git.chang-joon.lee@intel.com>
From: Hong Liu <hong.liu@intel.com>
Date: Fri, 24 Apr 2015 12:03:27 +0800
Subject: [PATCH 63/68] REVERTME [VPG]: drm/i915: get correct ddc bus number
 from vbt

Using ddc_pin field in VBT to initialize the ddc_bus number for hdmi
connecotr.

REVERTME: The VBT spec definition for this area is wrong for CHV, once
VBT spec is fixed, we need to submit a patch with a pointer to VBT spec
and definition.

Issue: IMINAN-26954
Change-Id: I64ae71b8112180c5c65052025b07c167d74fc807
Signed-off-by: Hong Liu <hong.liu@intel.com>
---
 drivers/gpu/drm/i915/intel_hdmi.c |   29 +++++++++++++++++++++++++----
 1 file changed, 25 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_hdmi.c b/drivers/gpu/drm/i915/intel_hdmi.c
index f7b0487..b2006de 100644
--- a/drivers/gpu/drm/i915/intel_hdmi.c
+++ b/drivers/gpu/drm/i915/intel_hdmi.c
@@ -1870,6 +1870,26 @@ intel_hdmi_add_properties(struct intel_hdmi *intel_hdmi, struct drm_connector *c
 	intel_hdmi->color_range_auto = true;
 }
 
+static void intel_hdmi_setup_ddc_vbt(struct drm_i915_private *dev_priv,
+				     struct intel_digital_port *intel_dig_port)
+{
+	struct intel_hdmi *intel_hdmi = &intel_dig_port->hdmi;
+	struct intel_encoder *intel_encoder = &intel_dig_port->base;
+	enum port port = intel_dig_port->port;
+	int i;
+
+	for (i = 0; i < dev_priv->vbt.child_dev_num; i++) {
+		if (port != dev_priv->vbt.child_dev[i].common.dvo_port)
+			continue;
+
+		intel_hdmi->ddc_bus = dev_priv->vbt.child_dev[i].common.ddc_pin;
+		break;
+	}
+
+	DRM_INFO("hdmi port %d, ddc_bus %d, hpd_pin %d\n",
+		 port, intel_hdmi->ddc_bus, intel_encoder->hpd_pin);
+}
+
 void intel_hdmi_init_connector(struct intel_digital_port *intel_dig_port,
 			       struct intel_connector *intel_connector)
 {
@@ -1898,10 +1918,7 @@ void intel_hdmi_init_connector(struct intel_digital_port *intel_dig_port,
 		intel_encoder->hpd_pin = HPD_PORT_C;
 		break;
 	case PORT_D:
-		if (IS_CHERRYVIEW(dev))
-			intel_hdmi->ddc_bus = GMBUS_PORT_DPD_CHV;
-		else
-			intel_hdmi->ddc_bus = GMBUS_PORT_DPD;
+		intel_hdmi->ddc_bus = GMBUS_PORT_DPD;
 		intel_encoder->hpd_pin = HPD_PORT_D;
 		break;
 	case PORT_A:
@@ -1911,6 +1928,10 @@ void intel_hdmi_init_connector(struct intel_digital_port *intel_dig_port,
 		BUG();
 	}
 
+	/* for cherryview, we prefer the ddc_bus number from VBT */
+	if (IS_CHERRYVIEW(dev))
+		intel_hdmi_setup_ddc_vbt(dev_priv, intel_dig_port);
+
 	if (IS_VALLEYVIEW(dev)) {
 		intel_hdmi->write_infoframe = vlv_write_infoframe;
 		intel_hdmi->set_infoframes = vlv_set_infoframes;
-- 
1.7.9.5

